### DISCUSSION TITLE
Environment deployment at scheduled time

### DISCUSSION THREAD

**Discussion by brandonmnewman** [community/community#28863] [Actions]
In GitHub Actions, through the use of environment protections you can delay a deployment by X minutes, but there doesn't seem to be a way to delay until X:XX PM. I have use cases where we want to approve a deployment into an environment during business hours but not actually run the deployment until a certain time after hours to prevent disrupting staff. Is there an accepted way to do this now or is it something that could be added? I looked at the road map and couldn't find anything that stood out. 

We would accomplish this in Azure DevOps Server via 'Stage Triggers' using 'Pre-deployment approvals' and a 'Schedule'. In Azure DevOps Server, Stages in a Release are roughly equivalent to a Job with an attached Environment in GitHub Actions and the 'Pre-deployment approvals' are roughly equivalent to 'Required Approvers'. The closest thing to a 'Stage Trigger Schedule' in GHA in the 'Wait Timer' but that is just a set number of minutes. Really it would be nice if that could include an ability to specify a schedule when it would run next, i.e. tonight at 7:00 PM. Including a link to the Azure DevOps docs on Stage Triggers so you can see what I am looking for.

https://docs.microsoft.com/en-us/azure/devops/pipelines/release/triggers?view=azure-devops

**Comment by GuillermoFarias** [comment]
Although I like the idea of ​​specifying what time to run a deployment in GHA, I think it's wrong.
Possible solutions according to my experience:

### GHActions and TZ validation
You can create an actions that runs every X minutes asking if there is a new tag and if the time according to TZ corresponds to deployment.
You can have an API that gives you information about deployment times.

### An external execute the deployment
An external create a tag in the repo using the GH API and have a CD flow in some tool like GHActions or AzureDevOps that runs the deployment.

**Reply by brandonmnewman** [reply]
It is a feature that exists in other CI/CD systems, and it seems reasonable to me that you might have maintenance windows outside business hours to limit impact on customers, so don't see how scheduling deployments is wrong.... It is also frustratingly close to being implemented already, just allow a time to be entered instead of a number of minutes.... If what you are suggesting with directing me back to Azure DevOps is that GitHub Actions is not supposed to be a fully featured CI/CD solution that is competitive in the space, that is disappointing as I have been putting a lot of effort into moving to this system and have been mostly pleased so far.

**Reply by GregJCook** [reply]
I agree with this response. I came to GitHub from Azure DevOps and this is definitely a feature I miss. We would review and approve our deployments in the afternoon, but delay them until after business hours. This allowed our release engineers to have their evenings back and the business to be happy to avoid mid-day downtime. 

Putting an arbitrary number of minutes to delay doesn't work because you are now boxed in to approving every release at the same time of the day to achieve the same time for the deployment to occur. It also blocks any hotfix deployments that need to circumvent the delay. Azure DevOps implemented this feature perfectly. You check a box to enable the delay and choose the date / time to execute the deployment.

**Reply by julealgon** [reply]
> Although I like the idea of ​​specifying what time to run a deployment in GHA, **I think it's wrong**.

Wait... what does this even mean? How can it be "wrong" to want to deploy something at a specified time?

This makes absolutely no sense to me.

Whether or not "scheduling a deployment" is right or wrong is completely subjective and up to each team to decide, not for the tool.

**Comment by tharren** [comment]
Yes, plugging in a time at the time of approval is a requirement for our teams as well. It finally came as a feature to Azure Devops yml pipeline but Github does not yet offer it

Azure Devops 'Defer Approval' functionality
![image](https://github.com/community/community/assets/7593374/185cc012-3b59-4ad6-a28d-465f60ea3135)
Github has no such feature
![image](https://github.com/community/community/assets/7593374/14d21eba-11ee-41e2-b3a8-3a447a3ebe62)

**Comment by julealgon** [comment]
It baffles me how such basic capabilities are _still_ missing from GHA _years_ after it was reported.

Why hasn't this been prioritized and implemented yet?

**Comment by cmeeren** [comment]
I wanted to migrate our company's pipelines from Azure Pipelines to GitHub Actions, but the lack of this specific requirement (and the code/infrastructure necessary to work around it) makes this a no-go for us.
